### DISCUSSION TITLE
Allow setting additional httpAgentOptions such as maxSockets & timeout

### DISCUSSION THREAD

**Discussion by longzheng** [vercel/next.js#43364] [Ideas]
### Describe the feature you'd like to request

Running a Next.js application (or any Node.js application) on Azure App Service may lead to SNAT (Source Network Address Translation) port exhaustion if there are a lot of outbound connections on the server-side, for example fetching API calls in `getServerSideProps`.

More in-depth explanation about App Service SNAT are available here:
https://learn.microsoft.com/en-us/azure/app-service/app-service-web-nodejs-best-practices-and-troubleshoot-guide#my-node-application-is-making-excessive-outbound-calls
https://azureossd.github.io/2022/03/10/NodeJS-with-Keep-Alives-and-Connection-Reuse/index.html

The easiest and most effective way to resolve this is to limit the maximum number of sockets opened by the application at any one time.

By default, [Next.js polyfills `node-fetch` for server-side `fetch` requests](https://nextjs.org/docs/basic-features/supported-browsers-features#server-side-polyfills). `node-fetch` supports setting a [custom agent with a Node.js `Agent`](https://github.com/node-fetch/node-fetch#custom-agent).

Currently Next.js supports specifying `httpAgentOptions` in `next.config.js`, but only to [enable/disable keep-alive](https://nextjs.org/docs/api-reference/next.config.js/disabling-http-keep-alive).

```javascript
module.exports = {
  httpAgentOptions: {
    keepAlive: false,
  },
}
```

### Describe the solution you'd like

`httpAgentOptions` should accept additional and probably all [`AgentOptions` properties](https://nodejs.org/api/http.html#new-agentoptions) including but not limited to `maxSockets` `maxFreeSockets` `timeout` so I can specify

```javascript
module.exports = {
    httpAgentOptions: {
        keepAlive: true,
        maxSockets: 128,
        maxFreeSockets: 128,
        timeout: 60000,
    },
}
```

The [type check should also be updated](https://github.com/vercel/next.js/blob/bcb0e282018045cba63090cc90451f4e5d08d3d7/packages/next/server/config-shared.ts#L419) so running the server with additional options will not warnings
```
[next] warn  - Invalid next.config.js options detected: 
[next]   - The value at .httpAgentOptions has an unexpected property, maxSockets, which is not in the list of allowed properties (keepAlive).
[next]   - The value at .httpAgentOptions has an unexpected property, maxFreeSockets, which is not in the list of allowed properties (keepAlive).
[next]   - The value at .httpAgentOptions has an unexpected property, timeout, which is not in the list of allowed properties (keepAlive).
```

However because the [entire `httpAgentOptions` object is passed along to `new HttpAgent()`](https://github.com/vercel/next.js/blob/bcb0e282018045cba63090cc90451f4e5d08d3d7/packages/next/server/config.ts#L87-L92) all the options still works.

### Describe alternatives you've considered

I investigated overriding the [global instance of `Agent` with `http.globalAgent`](https://nodejs.org/api/http.html#httpglobalagent) but because Next.js always [sets the agent to the custom version of `__NEXT_HTTP_AGENT`](https://github.com/vercel/next.js/blob/e650047d4b79d130292e75e89f43d127a2d8b4e2/packages/next/server/node-polyfill-fetch.js#L36-L39) it will never use the global instance.

**Comment by longzheng** [comment]
I also tried passing `agent` directly to my `fetch()` calls but it turns out the types do not allow me. https://github.com/vercel/next.js/issues/43354

**Comment by tjheffner** [comment]
This would be very useful.
